### PR TITLE
Store packets

### PR DESCRIPTION
**Pull request by dhruvja**
Store the packets in a vector so that it can be queried by the relayer. Right now we can store in the private storage and maybe later on, have a seperate PDA for the packets.

### CODE REVIEW

File: solana/solana-ibc/programs/solana-ibc/src/lib.rs [hunk 1]
```diff
@@ -122,6 +148,7 @@ pub type InnerClient = String; // Serialized
 pub type InnerConnectionEnd = String; // Serialized
 pub type InnerChannelEnd = String; // Serialized
 pub type InnerConsensusState = String; // Serialized
+pub type InnerPacketMsg = Vec<u8>;
```
Comment (blasrodri) [inline solana/solana-ibc/programs/solana-ibc/src/lib.rs:151]: why stored it serialized?

Comment (dhruvja) [reply solana/solana-ibc/programs/solana-ibc/src/lib.rs:151]: Because the `PacketMsg` doesnt implement `AnchorSerialize` and `AnchorDeserialize` macros

**Comment by dhruvja** [comment]
Storing the packets in a seperate PDA so that we dont run out of storage for our private storage.

File: solana/solana-ibc/programs/solana-ibc/src/lib.rs [hunk 2]
```diff
@@ -63,7 +67,29 @@ pub mod solana_ibc {
...
+                        {
+                            Some(_) => (),
+                            None => {
+                                inner_store.packets.0.push(serialized_packet)
+                            }
+                        }
+                    }
+                    _ => (),
+                }
+                result
```
Comment (mina86) [inline solana/solana-ibc/programs/solana-ibc/src/lib.rs:92]: Could we instead generate a new PDA on demand? Something with seed being `["packet", client_id, connection_id, port_id, channel_id, sequence]`.

Also, how are we deleting packets from the list?

File: solana/solana-ibc/programs/solana-ibc/src/lib.rs (hunk 1 above)
Comment (mina86) [reply solana/solana-ibc/programs/solana-ibc/src/lib.rs:153]: [AnchorSerialize and AnchorDeserialize are just aliases to Borsh](https://github.com/coral-xyz/anchor/blob/7e5bdce710f9580e1489089d04afb13f71a5d1c7/lang/src/lib.rs#L58) so PacketMsg does actually implement it.

File: solana/solana-ibc/programs/solana-ibc/src/lib.rs (hunk 2 above)
Comment (dhruvja) [reply solana/solana-ibc/programs/solana-ibc/src/lib.rs:92]: well we are not deleting the packets from the list. But i think we can delete them once they are acknowledged?

Comment (dhruvja) [reply solana/solana-ibc/programs/solana-ibc/src/lib.rs:92]: For PDA in demand, i think we can

**Review by blasrodri — APPROVED** [review]

File: solana/solana-ibc/programs/solana-ibc/src/lib.rs [hunk 3]
```diff
@@ -51,14 +55,33 @@ pub mod solana_ibc {
...
 
         let errors = messages
             .into_iter()
-            .map(|msg| ibc::core::dispatch(&mut store, &mut router, msg))
+            .map(|msg| {
+                let result =
+                    ibc::core::dispatch(&mut store, &mut router, msg.clone());
+                if let MsgEnvelope::Packet(packet) = msg {
+                    // store the packet if not exists
+                    // TODO(dhruvja) Store in a PDA with channelId, portId and Sequence
```
Comment (mina86) [inline solana/solana-ibc/programs/solana-ibc/src/lib.rs:71]: Long term we just need to get rid of storing packets in Solana.

**Review by mina86 — APPROVED** [review]

**Pull request merged**